### ISSUE TITLE
runUntilCurrent should check for the waker before trying to call wakeUp on it

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3611]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)| @therve reported|
|-|-|
|Trac ID|trac#3611|
|Type|defect|
|Created|2009-01-09 16:44:34Z|
|Branch|https://github.com/twisted/twisted/tree/wakeup-3611|

It's mainly a problem with IOCP which doesn't have a waker.

<details><summary>Searchable metadata</summary>

```
trac-id__3611 3611
type__defect defect
reporter__therve therve
priority__normal normal
milestone__Twisted_9_0 Twisted-9.0
branch__branches_wakeup_3611 branches/wakeup-3611
branch_author__therve therve
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1231519474000000 1231519474000000
changetime__1231749269000000 1231749269000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_1"></a>|@therve commented|
|-|-|

(In [25921]) Branching to 'wakeup-3611'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_2"></a>|@therve removed owner|
|-|-|

Please review! The changes are minimal, and hopefully not too much controversial.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun set owner to @therve|
|-|-|

I noticed that `ThreadedSelectReactor.wakeUp` is probably buggy and also probably useless ever since we we stopped checking which thread we were in in the base implementation of `wakeUp`.  We don't really have any tests for `ThreadedSelectReactor`, so it'd be hard to figure out if deleting that `wakeUp` is good or not.  Maybe it's worth filing another ticket for.

This change looks good, though, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve set status to `closed`|
|-|-|

(In [25943]) Merge wakeup-3611

Author: therve
Reviewer: exarkun
Fixes [#3611](https://github.com/twisted/twisted/issues/3611)

Check for the waker in runUntilCurrent when calling wakeUp. This fixes tests
for IOCP which doesn't have a waker.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
